cmd/deploy/input_test.go

```diff
@@ -2,9 +2,276 @@ package main
 
 import (
 	"encoding/json"
+	"maps"
+	"slices"
+	"strings"
 	"testing"
+
+	"cuelabs.dev/go/oci/ociregistry"
+	"github.com/docker-library/meta-scripts/registry"
 )
 
+func TestNormalizeInputRefs(t *testing.T) {
+	for _, x := range []struct {
+		name       string
+		deployType deployType
+		rawRefs    []string
+		wantRefs   []registry.Reference
+		wantDigest ociregistry.Digest
+		wantErr    string
+	}{
+		{
+			name:       "happy path",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"localhost:5000/foo@sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				"localhost:5000/bar:some-tag",
+				"localhost:5000/baz",
+			},
+			wantRefs: []registry.Reference{
+				{
+					Host:       "localhost:5000",
+					Repository: "foo",
+					Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				},
+				{
+					Host:       "localhost:5000",
+					Repository: "bar",
+					Tag:        "some-tag",
+				},
+				{
+					Host:       "localhost:5000",
+					Repository: "baz",
+				},
+			},
+			wantDigest: "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			wantErr:    "",
+		},
+		{
+			name:       "invalid digest",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"localhost:5000/foo@bad_digest",
+			},
+			wantErr: "failed to parse ref:",
+		},
+		{
+			name:       "invalid host",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"...",
+			},
+			wantErr: "failed to parse ref:",
+		},
+		{
+			name:       "invalid tag",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"localhost:5000/foo:#",
+			},
+			wantErr: "failed to parse ref:",
+		},
+		{
+			name:       "invalid reference url",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"localhost:5000/foo?test",
+			},
+			wantErr: "failed to parse ref:",
+		},
+		{
+			name:       "mismatch digest",
+			deployType: typeManifest,
+			rawRefs: []string{
+				"localhost:5000/foo@sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				"localhost:5000/foo@sha256:00042f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			wantErr: "refs digest mismatch in ",
+		},
+		{
+			name:       "can't push to blob",
+			deployType: typeBlob,
+			rawRefs: []string{
+				"localhost:5000/foo:tag",
+			},
+			wantErr: "cannot push blobs to a tag:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			refs, refsDigest, err := normalizeInputRefs(x.deployType, x.rawRefs)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match.\ngot:\n%q,\n\nexpected to contain:\n%q", err, x.wantErr)
```

Comment:
Perhaps https://github.com/stretchr/testify would make all these error reports simpler.

Reply:
@whalelines I kind of agree but I started a thread on testing frameworks in one of our channel and it seems go developers tend to not want to use such framework. I will let @tianon decide

Reply:
The signing team is using testify to good effect.

Reply:
In most cases, less is code is better. The less code, the less to maintain. The best refactor is delete.

Reply:
im all for it if @tianon and @yosifkit are good with it. Although, could we perhaps merge first and refactor after? Lots of activity in this PR already

Reply:
"A little copying is better than a little dependency." (https://go-proverbs.github.io/)

In this case, I don't think testify counts as a "little" dependency, but that's also often a good reason to avoid it -- generally in Go, dependencies are heavy, and create extra load on consumers of your code too.

I'm not opposed to `testify`, but the amount it makes our tests nicer needs to be greater than the amount of burden it (and its dependencies) place on us.  It looks like the dep tree for testify itself is pretty small, so it's probably reasonable, but I agree it's something to consider as a separate PR.